### DISCUSSION TITLE
Non Aligned Reads

### DISCUSSION THREAD

**Discussion by Hemanth-Eduru** [axboe/liburing#936] [Q&A] [answered]
Hi I am new to use of *liburing* I have the following doubts 
 1. What is the need of `posix_memalign` of the buffer which we will be passing to `io_uring_prep_read` I tried without alignment then it will throw error during *read* with `cqe failed: Invalid argument`
 2. Is it possible to read data which not a multiple of 4k Bytes (or power of 2)?

**Comment by axboe** [comment, marked as answer]
This isn't related to io_uring, these are O_DIRECT restrictions. For O_DIRECT, you have to read in multiples of the sector size of the device. Depending on the kernel version, the buffer used must also be aligned to the sector size, or (in newer kernels), just aligned to the DMA restrictions of the device.

**Reply by Hemanth-Eduru** [reply]
Thanks for the quick reply, one more doubt, are these applicable even if ***do not*** use O_DIRECT flag while opening the file (just O_RDONLY flag)

**Reply by axboe** [reply]
No, as mentioned, these are O_DIRECT only restrictions.

**Reply by Hemanth-Eduru** [reply]
Thank you
